File: java/src/main/java/ru/kontur/courses/chess/BoardParser.java
package ru.kontur.courses.chess;

import java.util.Arrays;

public class BoardParser {

    public Board parseBoard(String[] lines) throws Exception {
        if (lines.length != 8) throw new IllegalArgumentException("Should be exactly 8 lines");
        if (Arrays.stream(lines).anyMatch(it -> it.length() != 8))
            throw new IllegalArgumentException("All lines should have 8 chars length");

        var cells = new Piece[8][];
        for (var y = 0; y < 8; y++) {
            var line = lines[y];
            if (line == null) throw new Exception("incorrect input");
            cells[y] = new Piece[8];
            for (var x = 0; x < 8; x++)
                cells[y][x] = parsePiece(line.charAt(x));
        }
        return new Board(cells);
    }

    private static Piece parsePiece(char pieceSign) {
        var color = Character.isUpperCase(pieceSign) ? PieceColor.WHITE : PieceColor.BLACK;
        var pieceType = parsePieceType(Character.toUpperCase(pieceSign));
        return pieceType == null ? null : new Piece(pieceType, color);
    }

    private static PieceType parsePieceType(char sign) {
        switch (sign) {
            case 'R':
                return PieceType.rook;
            case 'K':
                return PieceType.king;
            case 'N':
                return PieceType.knight;
            case 'B':
                return PieceType.bishop;
            case 'Q':
                return PieceType.queen;
            case ' ':
            case '.':
                return null;
            default:
                throw new IllegalArgumentException("Unknown chess piece " + sign);
        }
    }
}

File: python/Chess/Solved/board_parser.py
import re
import piece_type
from board import Board
from piece import Piece
from piece_color import PieceColor

from typing import List, Union, NoReturn


class BoardParser:
    def parse_board(self, lines: List[List[str]]) -> Union[Board, NoReturn]:
        if len(lines) != 8:
            raise Exception("Should be exactly 8 lines")

        if any([len(line) != 8 for line in lines]):
            raise Exception("All lines should have 8 chars length")

        cells = []
        for line in lines:
            if not line:
                raise Exception("incorrect input")
            parsed_line = []
            for piece in line:
                parsed_line.append(self.parse_piece(piece))
            cells.append(parsed_line)

        return Board(cells)

    def parse_piece(self, piece_sign: str):
        color = (
            PieceColor.WHITE if re.search(r"[A-Z]", piece_sign) else PieceColor.BLACK
        )
        piece_type = self.parse_piece_type(piece_sign.upper())
        return Piece(piece_type, color) if piece_type else None

    @staticmethod
    def parse_piece_type(sign: str):
        sign_mapping = {
            "R": piece_type.Rook,
            "K": piece_type.King,
            "N": piece_type.Knight,
            "B": piece_type.Bishop,
            "Q": piece_type.Queen,
            " ": None,
            ".": None,
        }
        try:
            return sign_mapping[sign]
        except KeyError:
            raise Exception(f"Unknown chess piece {sign}")
